code/game/machinery/_machines_base/stock_parts/access_lock.dm

```diff
@@ -46,35 +57,101 @@
 /obj/item/stock_parts/access_lock/attackby(obj/item/W, mob/user)
 	var/obj/machinery/machine = loc
 	if(!emagged && istype(machine))
-		if(check_access(W))
+		var/obj/item/card/id/I = W.GetIdCard()
+		if(I && check_access(I))
 			locked = !locked
 			visible_message(SPAN_NOTICE("\The [src] beeps and flashes green twice: it is now [locked ? "" : "un"]locked."))
 			return TRUE
 		return
+	return ..()
 
-	if(isMultitool(W)) // This is a non-machine interaction only; doing this within a machine will likely lead to interaction clashes.
-		var/list/accesses = user.GetAccess()
-		var/list/choices = list()
-		for(var/key in accesses)
-			choices[get_access_desc(key)] = key
-		choices["No Access"] = "No Access"
-		var/default = length(req_access) ? req_access[1] : null
-		var/selected_desc = input(user, "Select the access you would like \the [src] to use:", "Access Modification", default) as null|anything in choices
-		. = TRUE
-		if(!CanPhysicallyInteract(user))
-			return
-		if(!selected_desc)
-			return
-		if(selected_desc == "No Access")
-			req_access.Cut()
-			return
-		if(!(choices[selected_desc] in user.GetAccess()))
-			return
-		req_access = list(choices[selected_desc])
-		return
+/obj/item/stock_parts/access_lock/attack_self(mob/user)
+	if (!ishuman(user) && !istype(user,/mob/living/silicon/robot))
+		return ..(user)
 
-	return ..()
+	ui_interact(user)
+
+/obj/item/stock_parts/access_lock/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = 1, datum/topic_state/state = GLOB.hands_state)
+	var/list/data = ui_data()
 
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "airlock_electronics.tmpl", src.name, 1000, 500, null, null, state)
+		ui.set_initial_data(data)
+		ui.open()
+
+/obj/item/stock_parts/access_lock/ui_data()
+	var/list/data = list()
+	var/list/regions = list()
+	if(!autoset)
+		for(var/i in ACCESS_REGION_MIN to ACCESS_REGION_MAX) //code/game/jobs/_access_defs.dm
+			var/list/region = list()
+			var/list/accesses = list()
+			for(var/j in get_region_accesses(i))
+				var/list/access = list()
+				access["name"] = get_access_desc(j)
+				access["id"] = j
+				access["req"] = conf_access && (j in conf_access)
+				accesses[++accesses.len] = access
+			region["name"] = get_region_accesses_name(i)
+			region["accesses"] = accesses
+			regions[++regions.len] = region
+		data["regions"] = regions
+		data["oneAccess"] = one_access
+	data["locked"] = locked
+	data["lockable"] = !emagged
+	data["autoset"] = autoset
+	return data
+
+/obj/item/stock_parts/access_lock/OnTopic(mob/user, list/href_list, state)
+	if(!emagged)
+		if(href_list["unlock"])
+			if(!req_access || istype(user, /mob/living/silicon))
```

Comment:
Why?

Reply:
ancestors must've really hated synths (this and comment above are on code i just staright up copyported from airlock electronics to lock)